From 53510f872ac468d3a1d4f3c1d0dd4e42428a7c1a Mon Sep 17 00:00:00 2001
From: lly <lly@sf.net>
Date: Fri, 10 Dec 2010 03:52:03 -0500
Subject: [PATCH] pppd: Debian patches

Using rp-pppoe pppd exits with EXIT_OK after receiving a timeout waiting for PADO due to no modem attached
(http://ppp.samba.org/cgi-bin/ppp-bugs/incoming?id=2211)

Add "lcp-echo-adaptive" option

Add support for the Framed-MTU Radius attribute
(http://ppp.samba.org/cgi-bin/ppp-bugs/incoming?id=1532)
---

--- pppd/pppd/lcp.c
+++ pppd/pppd/lcp.c
@@ -75,6 +75,7 @@ void saveWANStatus(char *currentstatus, int statusindex);
  */
 int	lcp_echo_interval = 0; 	/* Interval between LCP echo-requests */
 int	lcp_echo_fails = 0;	/* Tolerance to unanswered echo-requests */
+bool	lcp_echo_adaptive = 0;	/* request echo only if the link was idle */
 bool	lax_recv = 0;		/* accept control chars in asyncmap */
 bool	noendpoint = 0;		/* don't send/accept endpoint discriminator */
 
@@ -153,6 +154,8 @@ static option_t lcp_option_list[] = {
       OPT_PRIO },
     { "lcp-echo-interval", o_int, &lcp_echo_interval,
       "Set time in seconds between LCP echo requests", OPT_PRIO },
+    { "lcp-echo-adaptive", o_bool, &lcp_echo_adaptive,
+      "Suppress LCP echo requests if traffic was received", 1 },
     { "lcp-restart", o_int, &lcp_fsm[0].timeouttime,
       "Set time in seconds between LCP retransmissions", OPT_PRIO },
     { "lcp-max-terminate", o_int, &lcp_fsm[0].maxtermtransmits,
@@ -2345,6 +2348,22 @@ LcpSendEchoRequest (f)
     }
 
     /*
+     * If adaptive echos have been enabled, only send the echo request if
+     * no traffic was received since the last one.
+     */
+    if (lcp_echo_adaptive) {
+	static unsigned int last_pkts_in = 0;
+
+	update_link_stats(f->unit);
+	link_stats_valid = 0;
+
+	if (link_stats.pkts_in != last_pkts_in) {
+	    last_pkts_in = link_stats.pkts_in;
+	    return;
+	}
+    }
+
+    /*
      * Make and send the echo request frame.
      */
     if (f->state == OPENED) {
--- pppd/pppd/plugins/radius/radius.c
+++ pppd/pppd/plugins/radius/radius.c
@@ -651,6 +651,9 @@ radius_setparams(VALUE_PAIR *vp, char *msg, REQUEST_INFO *req_info,
 		    memcpy(rstate.class, vp->strvalue, rstate.class_len);
 		} /* else too big for our buffer - ignore it */
 		break;
+	    case PW_FRAMED_MTU:
+		netif_set_mtu(rstate.client_port,MIN(netif_get_mtu(rstate.client_port),vp->lvalue));
+		break;
 	    }
 
 
--- pppd/pppd/pppd.8
+++ pppd/pppd/pppd.8
@@ -570,6 +570,11 @@ to 1) if the \fIproxyarp\fR option is used, and will enable the
 dynamic IP address option (i.e. set /proc/sys/net/ipv4/ip_dynaddr to
 1) in demand mode if the local address changes.
 .TP
+.B lcp\-echo\-adaptive
+If this option is used with the \fIlcp\-echo\-failure\fR option then
+pppd will send LCP echo\-request frames only if no traffic was received
+from the peer since the last echo\-request was sent.
+.TP
 .B lcp\-echo\-failure \fIn
 If this option is given, pppd will presume the peer to be dead
 if \fIn\fR LCP echo\-requests are sent without receiving a valid LCP
-- 
2.10.5

